src/plugins/telemetry/server/telemetry_collection/get_data_telemetry/get_data_telemetry.ts

```diff
@@ -0,0 +1,244 @@
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { APICaller } from 'kibana/server';
+import { DATA_DATASETS_INDEX_PATTERNS, DataPatternName, DataTelemetryType } from './constants';
+
+export interface DataTelemetryBasePayload {
+  index_count: number;
+  ecs_index_count?: number;
+  doc_count?: number;
+  size_in_bytes?: number;
+}
+
+export interface DataTelemetryDocument extends DataTelemetryBasePayload {
+  dataset?: {
+    name?: string;
+    type?: DataTelemetryType | 'unknown' | string; // The union of types is to help autocompletion with some known `dataset.type`s
+  };
+  shipper?: string;
+  pattern_name?: DataPatternName;
+}
+
+export type DataTelemetryPayload = DataTelemetryDocument[];
+
+export interface DataTelemetryIndex {
+  name: string;
+  datasetName?: string; // To be obtained from `mappings.dataset.name` if it's a constant keyword
+  datasetType?: string; // To be obtained from `mappings.dataset.type` if it's a constant keyword
+  shipper?: string; // To be obtained from `_meta.beat` if it's set
+  isECS?: boolean; // Optional because it can't be obtained via Monitoring.
+
+  // The fields below are optional because we might not be able to obtain them if the user does not
+  // have access to the index.
+  docCount?: number;
+  sizeInBytes?: number;
+}
+
+type AtLeastOne<T, U = { [K in keyof T]: Pick<T, K> }> = Partial<T> & U[keyof U];
+
+type DataDescriptor = AtLeastOne<{
+  datasetName: string;
+  datasetType: string;
+  shipper: string;
+  patternName: DataPatternName; // When found from the list of the index patterns
+}>;
+
+function findMatchingDescriptors({
+  name,
+  shipper,
+  datasetName,
+  datasetType,
+}: DataTelemetryIndex): DataDescriptor | undefined {
+  // If we already have the data from the indices' mappings...
+  if ([shipper, datasetName, datasetType].some(Boolean)) {
+    return {
+      ...(shipper && { shipper }),
+      ...(datasetName && { datasetName }),
+      ...(datasetType && { datasetType }),
+    } as AtLeastOne<{ datasetName: string; datasetType: string; shipper: string }>; // Using casting here because TS doesn't infer at least one exists from the if clause
+  }
+
+  // Otherwise, try with the list of known index patterns
+  return DATA_DATASETS_INDEX_PATTERNS.find(({ pattern }) => {
```

Comment:
After a conversation with @alexfrancoeur and @kobelb, I need to change this to a `.filter`